Automattic/vip-go-mu-plugins
# Search: Keep all index versions in sync

## Description

Keep non-active index versions in sync when content is changed. This is accomplished in 3 ways:

* For deletes, the delete is processed on all indexes synchronously, as the queue system does not support deletes currently
* For regular content changes, we hook into `pre_ep_index_sync_queue` and send anything that is in the ElasticPress internal queue to our db-backed queue, to be processed asynchronously.
* For bulk changes or anything else that goes to our db-backed queue, we hook into a new action that fires when an item is queued, then replicate those jobs out to all the non-active index versions on shutdown.

NOTE - Something I realized - since we don't auto create the index when a new index version is added, after this PR we will end up with indexing jobs queued before the index is created. Need to figure out exactly what that means in practice.

## Checklist

Please make sure the items below have been covered before requesting a review:


## Steps to Test

1. Pull down PR
1. Update submodules (in `mu-plugins` in vip-go-mu-dev) - `git submodule update --init`
1. Add some new index versions (1 or 2) - `wp vip-search index-versions add post`
1. Open phpMyAdmin or similar and check the `wp_vip_search_index_queue` table - helps to clear it out too
1. Edit a post in WP Admin
1. Look at the table - there should be 1 job queued for the post on each index version you created
1. Clear out the table for simplicity
1. Add a new post and repeat - should have new jobs queued
1. Clear out the table
1. Run `wp vip-search queue stress-test`
1. Check the table - should have queued jobs for all posts on all index versions
1. Delete a post
1. Activate the new index (`wp vip-search index-versions activate post 2`) and do a site search (`/?s=test`) and make sure the post is not found via the new index (that it was deleted). Can also do this by hitting ES directly if that's easier.
